docs/user_guide/fairness_in_machine_learning.rst

```diff
@@ -141,6 +141,132 @@ which groups these values were observed, as well as the difference and ratio
 between the maximum and the minimum values. For more information refer to the
 subpackage :code:`fairlearn.metrics`.
 
+What traps can we fall into when abstracting a social problem?
+--------------------------------------------------------------
+
+Machine learning systems used in the real world are inherently sociotechnical
+systems, which include both technologies and social actors. Designers of machine
+learning systems who fail to consider the way in which social contexts and technologies
+are interrelated are at risk of falling into "abstraction traps". In this section,
+we explain what those traps are, and give some suggestions on how you can avoid them.
+
+In "Fairness and Abstraction in Sociotechnical Systems" Selbst et al. [#4]_
+identify failure modes that arise from applying abstraction. They identify
+them as:
+
+* *The Solutionism Trap*
+
+* *The Ripple Effect Trap*
+
+* *The Formalism Trap*
+
+* *The Portability Trap*
+
+* *The Framing Trap*
+
+Each of these traps are the result of failing to consider the way in
+with social context and technology are interrelated, as well as a deeper understanding of
+"the social" in order to solve problems Ackerman [#5]_.
+
+The Solutionism Trap
+^^^^^^^^^^^^^^^^^^^^
+
+This trap occurs when we assume that the best solution to a problem
+involves technology, and fail to recognize other possible solutions outside of
+this realm. One area where this manifests in is within contexts in which the
+definition "fairness" changes or is dependent on a political context.
+
+Example: consider the problem of internet connectivity in rural communities.
+An example of the solutionism trap is assuming that by using data science to
+study internet speed in a given region, policies and deals can be arranged to
+reduce the price of internet access. However, if there are additional socioeconomic
+problems within a community, for example lack of education, infrastructure, information
+technology and health services, then an algorithmic solution purely focused on internet
+speed will fail to meaningfully address the needs of the community.
+
+
+The Ripple Effect Trap
+^^^^^^^^^^^^^^^^^^^^^^
+
+Selbst et al. [#4]_ define this as a "failure to understand how the insertion of
+technology into an existing social system changes the behaviors and embedded values
+of the pre-existing system".
+
+Example 1:
+
+The Formalism Trap
+^^^^^^^^^^^^^^^^^^
+
+This trap occurs when implementing an algorithmic solution that fails to take
+into account the social dimensions associated to fairness in a situation. These
+dimensions include "procedurality", "contextuality" and "contestability". Such
+dimensions can often not be resolved via a purely mathematical framework.
```

Comment:
Maybe something that gets at _why_ they can't be resolved, like:

"Because different definitions of fairness cannot be satisfied at the same time (e.g., [Chouldechova et al., 2017](https://arxiv.org/pdf/1703.00056.pdf), [although feel free to replace with a better citation if people have other preferences]), you'll need to decide which definition to use. But there's no mathematical way to make that decision. And it might be that _none_ of the definitions are appropriate for your situation.

Some reasons why people fall into this trap is because fairness is contextually-dependent, open to contestation by different groups of people, and differences between ways of thinking about fairness between the legal world (i.e., fairness as procedural) and the fair-ML community (i.e., fairness as outcome-based)."

What do folks think? @lgfunderburk @hildeweerts

Reply:
This section took me a bit more to develop and was recently pushed. The reword and new examples look as follows:

`This trap occurs when abstraction results in failure to take
into account the full picture behind fairness in a situation. Many tasks of a data
scientist involve some form of formalization: from measuring real-world phenomena as
data to translating business KPI's and constraints into metrics, loss functions, or parameters.
We fall into the formalism trap when this translation fails to account the social dimensions
associated to the problem. Social dimensions that cannot be abstracted through a purely mathematical
framework include procedurality, contextuality and contestability.

Example: Consider the use of "risk scores" to decide the outcomes individuals
will experience. Risk scores may be used to determine risk of recidivism in crime,
risk of failure within a new role during recruitment, or risk of failure in a
patient's chosen treatment. A key element within these algorithms is the
ratio of false and true positives, as well as and false and true negatives.
That is, how successful an algorithm is to correctly asses risk in an individual.

A more important distinction is the impact of a wrongful assignment (is an individual
incarcerated, vs is an individual hired).

Take the following risk assessment formula from Kleinberg et al. [#6]_. The problem
of risk assessment can be abstracted as follows:

1. Feature vectors: this represent data we know about a given person.

2. Groups: this represents information regarding belonging of a person. Ideally our
algorithm does not display preference with respect to the group a person
belongs to. Example of groups include gender, race, age, etc.

3. Risk assignment: this represents the probability that a person with given
features, belonging to a given group, possesses a risk to engage in
well defined behaviour, and that the risk calculated accurately depicts reality for
that given person. In other words, the algorithm has low false positive rates and low false
negative rates.

From a mathematical point of view, it is possible to represent each of these
through the use of vectors, booleans (true or false) and probability values.
Where mathematical abstraction encounters a limitation is when capturing
information regarding procedurality (for example, how  do judges and police officers
determine whether bail, counselling, probation, or incarceration is appropriate);
contextuality (different societies determine what constitutes an "unfair" outcome, furthermore
different groups determine what constitutes immoral discrimination, i.e. is it immoral
to hire an applicant whose primary language is not English, for a non-English speaking role, over
an applicant whose only spoken language is English?); and contestability (the definitions
of discrimination and unfairness are politically contested and change over time, how
to mathematically represent definitions of unfairness as they are contested over time?).
`

It ended up being quite wordy. Open to reducing content

Reply:
For simplicity, you could start with some of that language I commented above:

"Because different definitions of fairness cannot be satisfied at the same time, you'll need to decide which definition to use. But there's no mathematical way to make that decision. And it might be that none of the definitions are appropriate for your situation.

Some reasons why people fall into this trap is because fairness is contextually-dependent, open to contestation by different groups of people, and differences between ways of thinking about fairness between the legal world (i.e., fairness as procedural) and the fair-ML community (i.e., fairness as outcome-based)."

Reply:
Adding this

Reply:
Ok, added